charger/mcc.go

```diff
@@ -0,0 +1,406 @@
+	// do we need to login?
+	if mcc.token == "" || time.Since(mcc.tokenValid) > 0 {
+		if err := mcc.login(mcc.password); err != nil {
+			return nil, err
+		}
+	}
+
+	// is it time to refresh the token?
+	if time.Since(mcc.tokenRefresh) > 0 {
+		if err := mcc.refresh(); err != nil {
+			return nil, err
+		}
+	}
+
+	// now lets process the request with the fetched token
+	req, err := http.NewRequest(method, uri, nil)
+	if err != nil {
+		return req, err
+	}
+
+	req.Header.Set("Authorization", fmt.Sprintf(" Bearer %s", mcc.token))
+
+	return req, nil
+}
+
+// use http PUT to set a value on the URI, the value should be URL encoded in the URI parameter
+func (mcc *MCC) putValue(uri string) error {
+	req, err := mcc.customRequest(http.MethodPut, uri)
+	if err != nil {
+		return err
+	}
+
+	b, err := mcc.Request(req)
+	if err == nil {
+		var result string
+		err = json.Unmarshal(b, &result)
+
+		if err == nil {
+			if result != "OK" {
+				return fmt.Errorf("Call returned an unexpected error")
+			}
+		}
+	}
+
+	return err
+}
+
+// use http GET to fetch a non structured value from an URI and stores it in result
+func (mcc *MCC) getValue(uri string, result interface{}) error {
+	req, err := mcc.customRequest(http.MethodGet, uri)
+	if err != nil {
+		return err
+	}
+
+	b, err := mcc.Request(req)
+	if err == nil {
+		err = json.Unmarshal(b, &result)
+	}
+
+	return err
+}
+
+// use http GET to fetch an escaped JSON string and unmarshall the data in result
+func (mcc *MCC) getEscapedJSON(uri string, result interface{}) error {
+	req, err := mcc.customRequest(http.MethodGet, uri)
+	if err != nil {
+		return err
+	}
+
+	b, err := mcc.Request(req)
+	if err == nil {
+		var unquote string
+		if err := json.Unmarshal([]byte(b), &unquote); err != nil {
+			return err
+		}
+
+		if err := json.Unmarshal([]byte(unquote), &result); err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+
+// Status implements the Charger.Status interface
+func (mcc *MCC) Status() (api.ChargeStatus, error) {
+	var chargeState int64
+
+	err := mcc.getValue(mcc.apiURL(apiChargeState), &chargeState)
+
+	if err != nil {
+		return api.StatusNone, err
+	}
+
+	// Charge State values and mapping to ChargeStatus
+	// 0: Unplugged		StatusA
+	// 1: Connecting	StatusB
+	// 2: Error				StatusE
+	// 3: Established	StatusF
+	// 4: Paused			StatusD
+	// 5: Active			StatusC
+	// 6: Finished		StatusD
+
+	switch chargeState {
+	case 0:
+		return api.StatusA, nil
+	case 1:
+		return api.StatusB, nil
+	case 2:
+		return api.StatusE, nil
+	case 3:
+		return api.StatusF, nil
+	case 4, 6:
+		return api.StatusD, nil
+	case 5:
+		return api.StatusC, nil
+	default:
+		return api.StatusNone, fmt.Errorf("properties unknown result: %d", chargeState)
+	}
+}
+
+// Enabled implements the Charger.Enabled interface
+func (mcc *MCC) Enabled() (bool, error) {
+	// Check if the car is connected and Paused, Active, or Finished
+	var chargeState int64
+
+	err := mcc.getValue(mcc.apiURL(apiChargeState), &chargeState)
```

Comment:
einzeiler?

Reply:
aber generell merkwürdig- unmarshal json in einen int? Ist das wirklich json und wie sieht das aus?

Reply:
Ne, das ist nun mal wirklich Blödsinn 😁 Da kommt einfach nur ein "String" zurück.